explainers/hidden-content-explainer.md

```diff
@@ -200,24 +205,87 @@ representing the active match is collapsed or it has a style which makes it
 invisible, the scroll will be canceled. Otherwise, the target match will be
 scrolled into view.
 
-### Alternatives Considered
+## Accessibility
+
+`content-visibility: hidden-matchable` subtrees will not be included in the
+accessibility tree because they are not visible on the screen. This is the same
+as `content-visibility: hidden` and offscreen/unrendered `content-visibility:
+auto` subtrees.
+
+The content inside `content-visibility: hidden-matchable` subtrees will still
+remain accessible to assisstive technology users because there should always be
+a way to reveal the content besides using find-in-page, such as a button to
+expand the content like a `<details>` element.
+
+## Scroll timing and element fragment issue
+
+With this feature, find-in-page and ScrollToTextFragment will have an
+asynchronous flow for firing the beforematch event and scrolling when the match
+is in a `content-visibility: hidden-matchable` subtree:
+1. A match within a `content-visibility: hidden-matchable` subtree has been
+   found
+2. Wait for the next requestAnimationFrame
+3. Fire the beforematch event on the `hidden-matchable` ancestor
+4. Wait for the next requestAnimationFrame
+5. Scroll to the target match
+
+This behavior was implemented in response to early feedback about pages where
+the beforematch event handler queued an asynchronous task to remove the
+`hidden-matchable` property, in which case the requestAnimationFrame gap between
+firing the beforematch event and scrolling was needed in order to correctly
+scroll to the revealed content.
+
+Supporting beforematch for element fragments is problematic because scrolling
+for element fragments is always done synchronously. To be more specific,
+consider the following code:
+```html
+<div style="height:100px"></div>
+<div id=foo></div>
+<script>
+window.scrollY; // returns 0
+window.location.hash = '#foo';
+window.scrollY; // returns 100
+</script>
+```
+Since changes to the element fragment trigger synchronous scrolling, which we
+can't change without a breaking the current behavior, we will have to fire the
+beforematch event and then scroll synchronously, instead of the
```

Comment:
I was thinking about this - why not just fire it synchronously like we do in the current implementation? I don't see how firing beforematch "at render timing" is more valuable than firing it before the scroll actually occurs, which we have the opportunity to do in this case...

Reply:
Firing events synchronously is a bad thing, because you have to be really careful to avoid re-entrancy of C++ code.
Another reason is that a synchronous event cannot be scheduled or throttled according to priorities. A third
is that an even that occurs synchronously during execution of another javascript task is bad, because it can result
in confusing code sequences for developers where code is expecting far from a callsite and they can't figure it
out.

Reply:
> you have to be really careful to avoid re-entrancy of C++ code

We could add a check for this before firing the beforematch event which would bail out or just not fire the beforematch event again if we are within a beforematch event handler, right? Or is that not the problem?

> a synchronous event cannot be scheduled or throttled according to priorities

For something like a scroll event handler, I could see why this is important in order to prevent jank, but when doing this once on load when navigating to a fragment, or when script itself changes the fragment, or when the user clicks on a link to change the fragment, I don't see how it would be janky compared to waiting for two frames like we did with find-in-page and ScrollToTextFragment. Or is this not about jank?

> an even that occurs synchronously during execution of another javascript task is bad, because it can result in confusing code sequences for developers where code is expecting far from a callsite and they can't figure it out.

Wouldn't it be more confusing for the beforematch event to actually happen after the match and after the scroll or not depending on what triggered the beforematch event (element fragment vs scroll to text fragment)? Would it make sense to make it async in the case where we are within another javascript task, which is a significantly less important use case than the c++ navigation one?

Reply:
I think it's fine to run the callback synchronously if this is not happening during another js task. However, I agree with chrishtr that we should not run a synchronous event handler in response to script changing window.location.hash, since that complicates what the developer may or may not assume about the state of the page during this function. As you mention, this should be a rare case, so I think that's fine. However, your example code above implies this and that should change.

Reply:
Anyway, I think we can just say "fire the event synchronously or via a posted task on the event loop" and then have the discussion about whether it's appropriate to fire it synchronously on the HTML spec issue.

Reply:
I added the either/or. How does it look?

Reply:
lgtm